src/passes/OptimizeInstructions.cpp

```diff
@@ -1995,39 +1995,60 @@ struct OptimizeInstructions
-    // if any of them has a type that cannot be cast to us, then we will trap,
-    // e.g.
-    //
-    //   (ref.cast $struct-A
-    //     (ref.cast $struct-B
-    //       (ref.cast $array
-    //         (local.get $x)
-    //
-    // The fallthrough is the local.get, but the array cast in the middle
-    // proves a trap must happen.
     Builder builder(*getModule());
-    auto nullType = curr->type.getHeapType().getBottom();
-    {
-      auto** refp = &curr->ref;
-      while (1) {
-        auto* ref = *refp;
 
-        auto result = GCTypeUtils::evaluateCastCheck(ref->type, curr->type);
+    // Look at all the fallthrough values to get the most precise possible type
+    // of the value we are casting. local.tee, br_if, and blocks can all "lose"
+    // type information, so looking at all the fallthrough values can give us a
+    // more precise type than is stored in the IR.
+    Type refType =
+      Properties::getFallthroughType(curr->ref, getPassOptions(), *getModule());
+
+    // If the cast admits null but we know the value is not null, then as a
+    // first step we can tighten up the cast to stop admitting nulls.
+    if (curr->type.isNullable() && refType.isNonNullable()) {
+      curr->type = Type(curr->type.getHeapType(), NonNullable);
+      // Call replaceCurrent() to make us re-optimize this node, as we may have
+      // just unlocked further opportunities. (We could just continue down to
+      // the rest, but we'd need to do more work to make sure all the local
+      // state in this function is in sync which this change; it's easier to
+      // just do another clean pass on this node.)
+      refinalize = true;
+      replaceCurrent(curr);
+      return;
+    }
 
-        if (result == GCTypeUtils::Success) {
-          // The cast will succeed. This can only happen if the ref is a subtype
-          // of the cast instruction, which means we can replace the cast with
-          // the ref.
-          assert(Type::isSubType(ref->type, curr->type));
-          if (curr->type != ref->type) {
-            refinalize = true;
-          }
-          // If there were no intermediate expressions, we can just skip the
-          // cast.
+    // Given what we know about the type of the value, determine what we know
+    // about the results of the cast and optimize accordingly.
+    switch (GCTypeUtils::evaluateCastCheck(refType, curr->type)) {
+      case GCTypeUtils::Unknown:
+        // The cast may or may not succeed, so we cannot optimize.
+        break;
+      case GCTypeUtils::Success:
+      case GCTypeUtils::SuccessOnlyIfNonNull: {
+        // We know the cast will succeed, so we can try to optimize it out. Find
+        // the best-typed fallthrough value to propagate.
+        auto** refp = Properties::getMostRefinedFallthrough(
+          &curr->ref, getPassOptions(), *getModule());
+        auto* ref = *refp;
+        assert(ref->type.isRef());
```

Comment:
Must this be a ref? What about
```wat
(ref.cast $foo
  (block (ref $foo)
    (unreachable)
  )
)
```
?

Reply:
Both `getFallthroughType` and `getMostRefinedFallthrough` return early if they encounter an unreachable, so they always produce referenced types and expressions with reference types as long as their original input had a reference type.

Reply:
But note that in my example the type of the block is *not* unreachable, it is `(ref $foo)`, even though the contents are unreachable. I see a check in the code for the type of the block but not for its content, so I think it will return the last item in the block in that case, which is unreachable.

Reply:
`getFallthroughType` and `getMostRefinedFallthrough` will return `(ref $foo)` and a pointer to the block, respectively here.

Reply:
I see, thanks, that's an API change compared to `getFallthrough` from before that confused me I guess.